### PR TITLE
8057543: Replace javac's Filter with Predicate (and lambdas)

### PR DESCRIPTION
**Pull request by lgxbslgx**
Hi all,

This patch replaces javac's `Filter` with `Predicate` and sets `Filter` as `Deprecated`.
Two existing tests failed and I fixed it.
Currently, all the tests in `test/langtools` passed locally by using the following command.

```
make test CONF=linux-x86_64-server-release JOBS=4 TEST=test/langtools/
```

Thank you for taking the time to review.

Best Regards.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8057543](https://bugs.openjdk.java.net/browse/JDK-8057543): Replace javac's Filter with Predicate (and lambdas)

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.java.net/census#mcimadamore) (@mcimadamore - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/1898/head:pull/1898` \
`$ git checkout pull/1898`

Update a local copy of the PR: \
`$ git checkout pull/1898` \
`$ git pull https://git.openjdk.java.net/jdk pull/1898/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1898`

View PR using the GUI difftool: \
`$ git pr show -t 1898`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/1898.diff">https://git.openjdk.java.net/jdk/pull/1898.diff</a>

</details>

### CODE REVIEW

**Comment by lgxbslgx** [comment]
More information and some similar issues:

These four issues is fixed in [this patch](https://github.com/openjdk/jdk/commit/c4e8276376676bda9aa4cb00b6d071fb70c21b85)
[JDK-8057542](https://bugs.openjdk.java.net/browse/JDK-8057542)
[JDK-8057545](https://bugs.openjdk.java.net/browse/JDK-8057545)
[JDK-8057546](https://bugs.openjdk.java.net/browse/JDK-8057546)
[JDK-8057547](https://bugs.openjdk.java.net/browse/JDK-8057547)

This issue was fixed in [this patch](https://github.com/openjdk/jdk/commit/582a4ebd92e60f95ae77d25eb00c6a146e727fab)
[JDK-8057544](https://bugs.openjdk.java.net/browse/JDK-8057544)

But their status is still `NEW`. It should be revised to `RESOLVED` or `CLOSE`.

**Review by mcimadamore — COMMENTED** [review]
This looks very good - sorry I missed it. If you reopen I will approve!

**Comment by lgxbslgx** [comment]
@mcimadamore Thank you for your help. I will reopen it.

**Comment by lgxbslgx** [comment]
I can't find the `reopen` button. Maybe it is my permission issue. Could anyone reopen it for me? Thanks a lot.

**Comment by lgxbslgx** [comment]
@mcimadamore Who should we ping to solve the `reopen` problem? I have several PRs which were closed by the robot and need to be reopen too.

**Comment by mcimadamore** [comment]
@edvbld Can you please help?

**Comment by lgxbslgx** [comment]
@mcimadamore The PR could be reopen by typing `/open` command. We can continue the work now. I would resolve the git conflicts later.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/api/JavacScope.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2006, 2014, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2006, 2020, Oracle and/or its affiliates. All rights reserved.
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/api/JavacScope.java:2]: copyright in this (and other files need to be updated)

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java [hunk 2]
```diff
@@ -90,7 +91,7 @@
 
     final Symtab syms;
     final JavacMessages messages;
-    final Names names;
+    private Names names;
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java:94]: where is this change coming from?

File: src/jdk.compiler/share/classes/com/sun/tools/javac/util/Filter.java [hunk 3]
```diff
@@ -28,7 +28,12 @@
 /**
  * Simple filter acting as a boolean predicate. Method accepts return true if
  * the supplied element matches against the filter.
+ *
+ * @deprecated
+ * It is better to use {@link java.util.function.Predicate}
+ * instead of {@link com.sun.tools.javac.util.Filter}.
  */
+@Deprecated(forRemoval = true, since = "17")
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/util/Filter.java:36]: This is not a public API - we're in a JDK internal class - if this class is no longer needed (as is the goal of this change) simply remove it :-)

**Review by mcimadamore — COMMENTED** [review]
Looks good - left some comments

File: src/jdk.compiler/share/classes/com/sun/tools/javac/util/Filter.java (hunk 3 above)
Comment (lgxbslgx) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/util/Filter.java]: Fixed.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java (hunk 2 above)
Comment (lgxbslgx) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java:94]: When building the JDK, I got the following error:

```
jdk/src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java:950: error: variable names might not have been initialized
                    t.name != names.init &&
                              ^
1 error
```

Therefore I revised the modifier of the `Names names`.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/api/JavacScope.java (hunk 1 above)
Comment (lgxbslgx) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/api/JavacScope.java]: Fixed.

**Comment by lgxbslgx** [comment]
> More information and some similar issues:
> 
> These four issues were fixed in [this patch](https://github.com/openjdk/jdk/commit/c4e8276376676bda9aa4cb00b6d071fb70c21b85)
> [JDK-8057542](https://bugs.openjdk.java.net/browse/JDK-8057542)
> [JDK-8057545](https://bugs.openjdk.java.net/browse/JDK-8057545)
> [JDK-8057546](https://bugs.openjdk.java.net/browse/JDK-8057546)
> [JDK-8057547](https://bugs.openjdk.java.net/browse/JDK-8057547)
> 
> This issue was fixed in [this patch](https://github.com/openjdk/jdk/commit/582a4ebd92e60f95ae77d25eb00c6a146e727fab)
> [JDK-8057544](https://bugs.openjdk.java.net/browse/JDK-8057544)
> 
> But their status are still `NEW`. It should be revised to `RESOLVED` or `CLOSED`.

@mcimadamore I want to set these issues to `CLOSED` or `RESOLVED`. What's your opinion?

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java (hunk 2 above)
Comment (forax) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java:94]: It's not a bug, it's a dubious feature, the lambda EG discussed that at length, you can ask Brian about it.

Usually replacing
```
t.name != names.init
```
by
```
  t.name != this.names.init
```
is enough.

**Comment by mcimadamore** [comment]
I agree that all these issues have been fixed - sorry - I thought these issues wanted to replace the various interfaces with interfaces from `j.u.function` (as done in this patch) - but these were about using lambda expressions more where possible - yeah I fixed that long ago.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java (hunk 2 above)
Comment (lgxbslgx) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java:94]: @forax Thank you for your reply. I tested `t.name != this.names.init` with `final Names names;` just now. The error also occurred. Same as `t.name != names.init`.

And I think that `t.name != names.init` and `t.name != this.names.init` are equal at this situation.

**Comment by lgxbslgx** [comment]
@mcimadamore OK, I closed them just now.

**Review by mcimadamore — APPROVED** [review]
Looks good!

**Comment by lgxbslgx** [comment]
@mcimadamore Thank you for your review. Could I get your help to sponsor this patch?
Although the Pre-submit tests don't finish, the tests on `tools/javac` passed locally.

/integrate

**Pull request closed without merging**